# Fibonacci Sequence Project

### 1. Compute the values for the fibonacci sequence: $$a_{5}, a_{6}, a_{7}, a_{8}$$ 

```Python
def fib(n):
    a = 0
    b = 1
    
    if n == 0:
        print(a)
        
    elif n == 1:
        print(b)
    
    else:
        print("n: 0, value: ", a) 
        print("n: 1, value: ", b) 
        for i in range(2, n):
            c = a + b
            a = b
            b = c
            if c <= 100:
                print("n: ", i, " value: ", c) 
fib(10)
```

#### $a_{5} = 5; $ $a_{6} = 8; $ $a_{7} = 13; $ $a_{8} = 21$

### 2. Write down the next few vectors $\vec{x_{3}},\vec{x_{4}},\vec{x_{5}},\vec{x_{6}},\vec{x_{7}}$

```Python

import numpy as np

def vectorFunc(n):
    a = 0
    b = 1
    vec = [[a],
           [b]]
    vector = np.array(vec)
    print("1: ", vector,"\n")
    for i in range (1, n):
        c = a + b
        a = b
        b = c
        vec = [[a],
               [b]]
        vector = np.array(vec)
        if c <= 100:
            print(i,": ",vector,"\n")
            
vectorFunc(100)
```       

$\vec{x_{3}} = \begin{bmatrix} 2 \\ 3 \end{bmatrix}, \vec{x_{4}} = \begin{bmatrix} 3 \\ 5 \end{bmatrix}, \vec{x_{5}} = \begin{bmatrix} 5 \\ 8\end{bmatrix}, \vec{x_{6}} = \begin{bmatrix} 8 \\ 13 \end{bmatrix}, \vec{x_{7}} = \begin{bmatrix} 13 \\ 21 \end{bmatrix}$  


### 3. Show that A computes Fibonacci numbers:

*A* = $\begin{bmatrix} 0 & 1 \\ 1 & 1 \end{bmatrix}$

#### (a) Compute the products *A*$\vec{x_{j}}$ for *j* = 1,2,3,4. What do you notice?

```Python
import numpy as np

def vectorMultiply(j):
    a = 0
    b = 1
    vec = [[a],
           [b]]
    A = [[0,1],[1,1]]
    vector = np.array(vec)
    matrixA = np.array(A)
    print("0: ", matrixA@vector,"\n")
    for i in range (1, j):
        c = a + b
        a = b
        b = c
        vec = [[a],
               [b]]
        vector = np.array(vec)
        result = matrixA@vector
        if c <= 100:
            print(i,": ",result,"\n")
            
vectorMultiply(100)
```

*A*$\vec{x_{1}} = \begin{bmatrix} 0 & 1 \\ 1 & 1 \end{bmatrix},$
*A*$\vec{x_{2}} = \begin{bmatrix} 0 & 1 \\ 2 & 2 \end{bmatrix},$
*A*$\vec{x_{3}} = \begin{bmatrix} 0 & 2 \\ 3 & 3 \end{bmatrix},$
*A*$\vec{x_{4}} = \begin{bmatrix} 0 & 3 \\ 5 & 5 \end{bmatrix},$


It is noteworthy that the last element in every matrix is the first element in every next matrix. The second row of matrix A has a dupicate of the upcoming fibonacci number, due to both columns of row 2 being 1.
This helps demonstrate that applying this matrix to one of the defined vectors gives the next vector in the sequence.

Known Fibonacci Sequence: 
  $\vec{x_{n}} =  \begin{bmatrix} \vec{a_{n}} \\ \vec{a_{n+1}}\end{bmatrix}$

$A\vec{x_{n}} = \begin{bmatrix} 0 & 1 \\ 1 & 1 \end{bmatrix}\begin{bmatrix} \vec{a_{n}} \\ \vec{a_{n+1}} \end{bmatrix}$
--> For n=1, $\vec{a_{1}}$ is 0. $\vec{a_{2}}$ can be assigned to $\vec{a_{n}}$:
This yields the vector:-->
$\begin{bmatrix} \vec 0 \\ \vec{a_{n}}\end{bmatrix}$
--> For n=2, $\vec{x_{2}} = \begin{bmatrix} \vec{a_{n}} \\ \vec{a_{n+1}}\end{bmatrix}$

-->This pattern continues for all real n:
$\vec{x_{n+1}} = \begin{bmatrix} \vec{a_{n+1}} \\ \vec{a_{n+2}}\end{bmatrix}$

#### C)
By Induction:

$\vec{y_{1}}$ = A$\vec{y_{0}}$ 

Base case

if $\vec{y_{n-1}}$ = $A{^{(n-1)}}$$\vec{y_{0}}$, we know that $A\vec{y_{n-1}}$ = $\vec{y_{n}}$

$\vec{y_{n-1}}$ = $A{^{(n-1)}}\vec{y_{0}}$

Therefore,
$\vec{y_{n}}$ = $A{^{n}}\vec{y_{0}}$

In [ ]:
import numpy as py
from numpy.linalg import eig
A = np.array([[0,1],[1,1]])
values, vectors = eig(A)
print("values:", values)
print("vectors:", vectors)

values: [-0.61803399  1.61803399]
vectors: [[-0.85065081 -0.52573111]
 [ 0.52573111 -0.85065081]]



### 4.
#### A)Eigenvalues of A
  $\lambda_{1} = \phi_{1}$     ;      $\lambda_{2} = \phi{_{2}}: $
  
  -0.61803399,           1.61803399
  
  (1-$\sqrt(5)$)/2  ;  (1+$\sqrt(5)$)/2

#### B) Non-zero Eigenvectors of A

$\begin{bmatrix} -\phi_{1}&0 \\ 0&\phi_{1} \end{bmatrix}$ $\begin{bmatrix} x \\ y \end{bmatrix}$ = 0   

$-\phi_{1}x$ + y = 0

y = $\phi_{1}$x

$\vec{v_{1}} = \begin{bmatrix} 1 \\ \phi_{1} \end{bmatrix}$ -> Same thing for $\vec{v_{2}}$ 
->
$\vec{v_{2}} = \begin{bmatrix} 1 \\ \phi_{2} \end{bmatrix}$

#### 5. Diagnolized Matrix

D = $\begin{bmatrix} \phi_{1}&0 \\ 0&\phi_{2} \end{bmatrix}$ 

P = $\begin{bmatrix} 1&1 \\ \phi_{1}&\phi_{2} \end{bmatrix}$ 

$P{^{-1}} = \frac{1}{(\phi{_{2}} - \phi{_{1}})} \begin{bmatrix} \phi_{2}&-1 \\ -\phi_{1}&1 \end{bmatrix}$ 


#### 6. Formula for $A{^{n}}$

$A{^{2}} = A * A = (PDP{^{-1}})(PDP{^{-1}})$ -> $PP{^{-1}} = 1$  
$DD = D{^{2}}$
= $PD{^{2}}P{^{-1}}$

Therefore,

$A{^{n}} = P\begin{bmatrix}\phi{_{1}}{^{n}}&0 \\ 0&\phi{_{2}}{^{n}}\end{bmatrix}P{^{-1}}$

$A{^{n}} = P\begin{bmatrix}\phi{_{1}}{^{n}}&0 \\ 0&\phi{_{2}}{^{n}}\end{bmatrix}P{^{-1}}$

$A{^{n}} = \begin{bmatrix} 1&1 \\ \phi{_{1}}&\phi{_{2}}\end{bmatrix}\begin{bmatrix}\phi{_{1}}{^{n}}&0 \\ 0&\phi{_{2}}{^{n}}\end{bmatrix}\begin{bmatrix}\phi{_{2}}&-1 \\ -\phi{_{1}}&1\end{bmatrix}$

$A{^{n}} = \begin{bmatrix}\phi{_{1}}{^{n}}\phi{_{2}} - \phi{_{2}}{^{n}}\phi{_{1}} & \phi{_{2}}{^{n}} - \phi{_{1}}{^{n}} \\\phi{_{2}}{^{n+1}} - \phi{_{2}} & \phi{_{2}}{^{n+1}}-\phi{_{1}}{^{n+1}}\end{bmatrix}\frac{1}{(\phi{_{2}} - \phi{_{1}})}$

$A{^{n}} = \begin{bmatrix}\phi{_{2}}{^{n-1}} - \phi{_{1}}{^{n-1}} & \phi{_{2}}{^{n}} - \phi{_{1}}{^{n}} \\\phi{_{2}}{^{n+1}} - \phi{_{2}} & \phi{_{2}}{^{n+1}}-\phi{_{1}}{^{n+1}}\end{bmatrix}\frac{1}{(\phi{_{2}} - \phi{_{1}})}$



#### 7. Find a formula for $\vec{a{^n}}$

$\vec{x_{n}} = A{^{n}}\vec{x_{0}}$

$\vec{x_{n}}$ = $\frac{1}{(\phi{_{2}} - \phi{_{1}})}$ $\begin{bmatrix}\phi{_{2}}{^{n-1}} - \phi{_{1}}{^{n-1}} & \phi{_{2}}{^{n}} - \phi{_{1}}{^{n}} \\ \phi{_{2}}{^{n+1}} - \phi{_{2}} & \phi{_{2}}{^{n+1}}-\phi{_{1}}{^{n+1}}\end{bmatrix}$

$\vec{x_{n}}$ = $\frac{1}{(\phi{_{2}} - \phi{_{1}})}$ $\begin{bmatrix}\phi{_{2}}{^{n}} - \phi{_{1}}{^{n}}\\ \phi{_{2}}{^{n+1}} - \phi{_{1}}{^{n+1}}\end{bmatrix}$ 

$\vec{x_{2}} = \begin{bmatrix}a{_{2}}\\ a{_{3}}\end{bmatrix}$. First row is $\vec{a_{n}}$

$\vec{x_{0}}$ = $\frac{1}{(\phi{_{2}} - \phi{_{1}})}$ $\begin{bmatrix}\phi{_{2}}{^{n-1}} - \phi{_{1}}{^{n-1}} & \phi{_{2}}{^{n}} - \phi{_{1}}{^{n}} \\ \phi{_{2}}{^{n+1}} - \phi{_{2}} & \phi{_{2}}{^{n+1}}-\phi{_{1}}{^{n+1}}\end{bmatrix}\begin{bmatrix} 0 \\ 1 \end{bmatrix}$

$\vec{x_{0}}$ = $\frac{1}{(\phi{_{2}} - \phi{_{1}})}$ $\begin{bmatrix} 0 \\ \phi{_{2}} - \phi{_{1}} \end{bmatrix}$ 

Check:

$\vec{x_{2}}$ = $\frac{1}{(\phi{_{2}} - \phi{_{1}})}$ $\begin{bmatrix} \phi{_{2}}{^{2}} - \phi{_{1}}{^{2}} \\ \phi{_{2}}{^{3}} - \phi{_{1}}{^{3}} \end{bmatrix}$ = $\begin{bmatrix} 1 \\ 2 \end{bmatrix}$

-- Therefore,

$\vec{a_{n}} = \frac{\phi{_{2}}{^{n}} - \phi{_{1}}{^{n}}}{\phi{_{2}} - \phi{_{1}}}$


#### 8. Making a good approximation for $\vec{a_{n}}$ using the fact that $lim{_{n->infinity}} = 0$

$lim{_{n->infinity}}\vec{a{_n}} = \frac{\phi{_{2}}{^{n}} - \phi{_{1}}{^{n}}}{\phi{_{2}} - \phi{_{1}}}$

For large values of n, $\phi{_{2}}{^{n}}$ goes to 0, so the approximation is

$\vec{a{_n}}$ is approximately equal to $\frac{\phi{_{1}}{^{n}}}{\phi{_{2}} - \phi{_{1}}}$ for large values of n

# Pell Numbers

## 1.
```Python
def pell(n):
    a = 0
    b = 1
    
    if n == 0:
        print(a)
        
    elif n == 1:
        print(b)
    
    else:
        print("n: 0, value: ", a) 
        print("n: 1, value: ", b) 
        for i in range(2, n):
            c = a + 2*b
            a = b
            b = c
            if c <= 1000:
                print("n: ", i, " value: ", c) 
pell(10)
```

#### $a_{5} = 29; $ $a_{6} = 70; $ $a_{7} = 169; $ $a_{8} = 408 $

```Python
import numpy as np

def vectorFunc(n):
    a = 0
    b = 1
    vec = [[a],
           [b]]
    vector = np.array(vec)
    print("1: ", vector,"\n")
    for i in range (1, n):
        c = a + 2*b
        a = b
        b = c
        vec = [[a],
               [b]]
        vector = np.array(vec)
        if c <= 1000:
            print(i,": ",vector,"\n")
            
vectorFunc(100)
```

$\vec{x_{3}} = \begin{bmatrix}5\\12\end{bmatrix}$
$\vec{x_{4}} = \begin{bmatrix}12\\29\end{bmatrix}$ 
$\vec{x_{5}} = \begin{bmatrix}29\\70\end{bmatrix}$ 
$\vec{x_{6}} = \begin{bmatrix}70\\169\end{bmatrix}$ 
$\vec{x_{7}} = \begin{bmatrix}169\\408\end{bmatrix}$ 

## 3.
```Python
def vectorMultiply(j):
    a = 0
    b = 1
    vec = [[a],
           [b]]
    B = [[0,1],[1,1]]
    vector = np.array(vec)
    matrixA = np.array(A)
    print("0: ", matrixA@vector,"\n")
    for i in range (1, j):
        c = a + 2*b
        a = b
        b = c
        vec = [[a],
               [b]]
        vector = np.array(vec)
        result = matrixA@vector
        if c <= 1000:
            print(i,": ",result,"\n")
            
vectorMultiply(100)
``` 

### a)
$\vec{x_{1}} = \begin{bmatrix}12\\17\end{bmatrix}$ 
$\vec{x_{2}} = \begin{bmatrix}29\\41\end{bmatrix}$ 
$\vec{x_{3}} = \begin{bmatrix}70\\99\end{bmatrix}$ 
$\vec{x_{4}} = \begin{bmatrix}169\\239\end{bmatrix}$ 

### b)
Known Pell Sequence: 
$\vec{x_{n}} =  \begin{bmatrix} \vec{a_{n}} \\ \vec{2a_{n+1}}\end{bmatrix}$
  
$\vec{x_{n+1}} = B\vec{x{_n}}$ 

$\begin{bmatrix} \vec{a_{n+1}} \\ \vec{a_{n+2}}\end{bmatrix}$ = $\begin{bmatrix} b&c \\ d&e\end{bmatrix}$$\begin{bmatrix} \vec{a_{n}} \\ \vec{2a_{n+1}}\end{bmatrix}$

$\vec{a_{n+1}}$ =  $b\vec{a_{n}}$ + $c\vec{a_{n+1}}$

b = 0; c = 1

$\vec{a_{n+2}}$ =  $d\vec{a_{n}}$ + $e\vec{2a_{n+1}}$

d = 1; e = 2

Therefore, matrix B:

$\begin{bmatrix} 0&1 \\ 1&2\end{bmatrix}$

--> For n=1, $\vec{a_{1}}$ is 0. $\vec{a_{2}}$ can be assigned to $\vec{a_{n}}$:
This yields the vector:-->
$\begin{bmatrix} \vec 0 \\ \vec{2a_{n}}\end{bmatrix}$
--> For n=2, $\vec{x_{2}} = \begin{bmatrix} \vec{a_{n}} \\ \vec{2a_{n+1}}\end{bmatrix}$

-->This pattern continues for all real n:
$\vec{x_{n+1}} = \begin{bmatrix} \vec{a_{n+1}} \\ \vec{2a_{n+2}}\end{bmatrix}$

#### C)
By induction: 
$\vec{y_{1}}$ = $B\vec{y_{0}}$ 

Base case

if $\vec{y_{n-1}}$ = $B{^{(n-1)}}\vec{y_{0}}$, we know that $B\vec{y_{n-1}}$ = $\vec{y_{n}}$

$\vec{y_{n-1}}$ = $B{^{(n-1)}}\vec{y_{0}}$

Therefore,
$\vec{y_{n}}$ = $B{^{n}}\vec{y_{0}}$


In [ ]:
import numpy as py
from numpy.linalg import eig
B = np.array([[0,1],[1,2]])
values, vectors = eig(B)
print("values:", values)
print("vectors:", vectors)

values: [-0.41421356  2.41421356]
vectors: [[-0.92387953 -0.38268343]
 [ 0.38268343 -0.92387953]]



### 4.
#### A)Eigenvalues of B
  $\lambda_{1} = \omega_{1}=-0.41421356$ 
  $\lambda_{2} = \omega_{2}=2.41421356$   

#### B) Non-zero Eigenvectors of B

$\begin{bmatrix} -\omega_{1}&0 \\ 0&\omega_{1} \end{bmatrix}$ $\begin{bmatrix} x \\ y \end{bmatrix}$ = 0   

$-\omega_{1}x$ + y = 0

y = $\omega{_{1}}x$

$\vec{v_{1}} = \begin{bmatrix} 1 \\ \omega_{1} \end{bmatrix}$ -> Same thing for $\vec{v_{2}}$ 
->
$\vec{v_{2}} = \begin{bmatrix} 1 \\ \omega_{2} \end{bmatrix}$

#### 5. Diagnolized Matrix

D = $\begin{bmatrix} \omega_{1}&0 \\ 0&\omega_{2} \end{bmatrix}$ 

P = $\begin{bmatrix} 1&1 \\ \omega_{1}&\omega_{2} \end{bmatrix}$ 

$P{^{-1}} = \frac{1}{(\omega{_{2}} - \omega{_{1}})} \begin{bmatrix} \omega_{2}&-1 \\ -\omega_{1}&1 \end{bmatrix}$ 

#### 6. Formula for $B{^{n}}$

$B{^{2}} = B * B = (PDP{^{-1}})(PDP{^{-1}})$ -> $PP{^{-1}} = 1$  
$DD = D{^{2}}$
= $PD{^{2}}P{^{-1}}$

Therefore,

$B{^{n}} = P\begin{bmatrix}\omega{_{1}}{^{n}}&0 \\ 0&\omega{_{2}}{^{n}}\end{bmatrix}P{^{-1}}$

$B{^{n}} = P\begin{bmatrix}\omega{_{1}}{^{n}}&0 \\ 0&\omega{_{2}}{^{n}}\end{bmatrix}P{^{-1}}$

$B{^{n}} = \begin{bmatrix} 1&1 \\ \omega{_{1}}&\omega{_{2}}\end{bmatrix}\begin{bmatrix}\omega{_{1}}{^{n}}&0 \\ 0&\omega{_{2}}{^{n}}\end{bmatrix}\begin{bmatrix}\omega{_{2}}&-1 \\ -\omega{_{1}}&1\end{bmatrix}$

$B{^{n}} = \begin{bmatrix}\omega{_{1}}{^{n}}\omega{_{2}} - \omega{_{2}}{^{n}}\omega{_{1}} & \omega{_{2}}{^{n}} - \omega{_{1}}{^{n}} \\\omega{_{2}}{^{n+1}} - \omega{_{2}} & \omega{_{2}}{^{n+1}}-\omega{_{1}}{^{n+1}}\end{bmatrix}\frac{1}{(\omega{_{2}} - \omega{_{1}})}$

$B{^{n}} = \begin{bmatrix}\omega{_{2}}{^{n-1}} - \omega{_{1}}{^{n-1}} & \omega{_{2}}{^{n}} - \omega{_{1}}{^{n}} \\\omega{_{2}}{^{n+1}} - \omega{_{2}} & \omega{_{2}}{^{n+1}}-\omega{_{1}}{^{n+1}}\end{bmatrix}\frac{1}{(\omega{_{2}} - \omega{_{1}})}$


#### 7. Find a formula for $\vec{a{^n}}$

$\vec{x_{n}} = A{^{n}}\vec{x_{0}}$

$\vec{x_{n}}$ = $\frac{1}{(\omega{_{2}} - \omega{_{1}})}$ $\begin{bmatrix}\omega{_{2}}{^{n-1}} - \omega{_{1}}{^{n-1}} & \omega{_{2}}{^{n}} - \omega{_{1}}{^{n}} \\ \omega{_{2}}{^{n+1}} - \omega{_{2}} & \omega{_{2}}{^{n+1}}-\omega{_{1}}{^{n+1}}\end{bmatrix}$

$\vec{x_{n}}$ = $\frac{1}{(\omega{_{2}} - \omega{_{1}})}$ $\begin{bmatrix}\phi{_{2}}{^{n}} - \phi{_{1}}{^{n}}\\ \omega{_{2}}{^{n+1}} - \omega{_{1}}{^{n+1}}\end{bmatrix}$ 

$\vec{x_{2}} = \begin{bmatrix}a{_{2}}\\ a{_{3}}\end{bmatrix}$. First row is $\vec{a_{n}}$

$\vec{x_{0}}$ = $\frac{1}{(\omega{_{2}} - \omega{_{1}})}$ $\begin{bmatrix}\omega{_{2}}{^{n-1}} - \omega{_{1}}{^{n-1}} & \omega{_{2}}{^{n}} - \omega{_{1}}{^{n}} \\ \omega{_{2}}{^{n+1}} - \omega{_{2}} & \omega{_{2}}{^{n+1}}-\omega{_{1}}{^{n+1}}\end{bmatrix}\begin{bmatrix} 0 \\ 1 \end{bmatrix}$

$\vec{x_{0}}$ = $\frac{1}{(\omega{_{2}} - \omega{_{1}})}$ $\begin{bmatrix} 0 \\ \omega{_{2}} - \omega{_{1}} \end{bmatrix}$ 

Check:

$\vec{x_{2}}$ = $\frac{1}{(\omega{_{2}} - \omega{_{1}})}$ $\begin{bmatrix} \omega{_{2}}{^{2}} - \omega{_{1}}{^{2}} \\ \omega{_{2}}{^{3}} - \omega{_{1}}{^{3}} \end{bmatrix}$ = $\begin{bmatrix} 1 \\ 2 \end{bmatrix}$

-- Therefore,
$\vec{a_{n}} = \frac{\omega{_{2}}{^{n}} - \omega{_{1}}{^{n}}}{\omega{_{2}} - \omega{_{1}}}$

#### 8. Making a good approximation for $\vec{a_{n}}$ using the fact that $lim{_{n->infinity}} = 0$

$lim{_{n->infinity}}\vec{a{_n}} = \frac{\omega{_{2}}{^{n}} - \omega{_{1}}{^{n}}}{\omega{_{2}} - \omega{_{1}}}$

For large values of n, $\omega{_{2}}{^{n}}$ goes to 0, so the approximation is

$\vec{a{_n}}$ is approximately equal to $\frac{\omega{_{1}}{^{n}}}{\omega{_{2}} - \omega{_{1}}}$ for large values of n